File: cpp/105/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildTree(preorder, 0, preorder.size()-1, inorder, 0, inorder.size()-1);
    }

    TreeNode* buildTree(vector<int> &preorder, int lp, int rp, vector<int> &inorder, int li, int ri){
        if(lp>rp) return NULL;
        TreeNode *root = new TreeNode(preorder[lp]);
        int i = li;
        while(inorder[i]!=preorder[lp])
            i++;
        root->left = buildTree(preorder, lp+1, lp+i-li, inorder, li, i-1);
        root->right = buildTree(preorder, lp+i-li+1, rp, inorder, i+1, ri);
        return root;
    }
};

File: python/105/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if not inorder or not preorder:
            return None
        root = TreeNode(preorder[0])
        position = inorder.index(preorder[0])
        root.left = self.buildTree(preorder[1:position+1], inorder[:position])
        root.right = self.buildTree(preorder[position+1:], inorder[position+1:])
        return root
